### PR TITLE
Join Graphs (after #446)

### PR DESCRIPTION
**Pull request by mrzzzrm**
Introduces the `JoinGraph`, which is the foundation of any JoinOrdering algorithm.

**Note:** Currently branched off from #446, will rebase once that is merged. For review convenience, the base until then is `feature/map_column_ids`

(@mrks this already includes the functionality of detecting "JoinPredicates", i.e. predicates that could be turned into a Join. Thus it is the first step towards replacing the malfunctioning JoinDetectionRule).

### CODE REVIEW

File: src/lib/optimizer/join_graph.hpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  ColumnID column_id;
+  ScanType scan_type;
+  AllParameterVariant value;
+};
+
+struct JoinVertex {
+  explicit JoinVertex(const std::shared_ptr<AbstractASTNode>& node);
+
+  std::shared_ptr<AbstractASTNode> node;
+  std::vector<JoinVertexPredicate> predicates;
```
Comment (Bensk1) [inline src/lib/optimizer/join_graph.hpp]: When do we use the word `predicate` and when do we use `edge`?

Does a `JoinVertex` connect always two `JoinNodes`? What is the maximum number of `predicates` then?

File: src/lib/optimizer/join_graph.hpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  std::pair<JoinVertexID, JoinVertexID> vertex_indices;
+  std::optional<std::pair<ColumnID, ColumnID>> column_ids;
+  JoinMode join_mode;
+  std::optional<ScanType> scan_type;
+};
+
+struct JoinVertexPredicate {
+  JoinVertexPredicate(ColumnID column_id, ScanType scan_type, const AllParameterVariant& value);
+
+  ColumnID column_id;
```
Comment (Bensk1) [inline src/lib/optimizer/join_graph.hpp]: Are these members going to change their value or could we make them `const`?

File: src/lib/optimizer/join_graph.cpp [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    // Everything that is not a Join or a Predicate becomes a vertex
+    o_vertices.emplace_back(JoinVertex(node));
+    return;
+  }
+}
+
+void JoinGraph::_traverse_inner_join_node(const std::shared_ptr<JoinNode>& node, JoinGraph::Vertices& o_vertices,
+                                          JoinGraph::Edges& o_edges) {
+  // The first vertex in o_vertices that belongs to the left subtree.
+  const auto first_left_subtree_vertex_idx = make_join_vertex_id(o_vertices.size());
```
Comment (Bensk1) [inline src/lib/optimizer/join_graph.cpp]: Accessing `o_vertices.size()` sounds like accessing the last vertex, not the first.

Comment (Bensk1) [reply src/lib/optimizer/join_graph.cpp]: In addition, `first_right_subtree_vertex_idx` in line 116 should contain the exact same value. In combination with the comment, this should confuse the reader

File: src/lib/optimizer/join_graph.cpp [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+   * For example: we found an AST node that we want to turn into a JoinEdge. The AST node is referring to two
+   * ColumnIDs, one in the left subtree and one in the right subtree. We need to figure out which vertices it is
+   * actually referring to, in order to form an edge.
+   *
+   * Think of the following table being generated by the left subtree:
+   *
+   * 0   | 1   | 2   | 3   | 4   | 5
+   * a.a | a.b | a.c | b.a | c.a | c.b
+   *
+   * Now, if the join_column_ids.left is "4" it is actually referring to vertex "c" (with JoinVertexID "2") and
```
Comment (Bensk1) [inline src/lib/optimizer/join_graph.cpp]: Should we add the `JoinVertexID` to the example above?

File: src/lib/optimizer/join_graph.cpp [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+JoinGraph::JoinGraph(Vertices&& vertices, Edges&& edges) : _vertices(std::move(vertices)), _edges(std::move(edges)) {}
+
+const JoinGraph::Vertices& JoinGraph::vertices() const { return _vertices; }
+
+const JoinGraph::Edges& JoinGraph::edges() const { return _edges; }
+
+void JoinGraph::print(std::ostream& out) const {
+  out << "==== JoinGraph ====" << std::endl;
+  out << "==== Vertices ====" << std::endl;
+  for (size_t vertex_idx = 0; vertex_idx < _vertices.size(); ++vertex_idx) {
```
Comment (Bensk1) [inline src/lib/optimizer/join_graph.cpp]: Since you added `JoinVertexID` as data type, we should either use it here or use `auto`

**Comment by mrzzzrm** [comment]
Ask discussed today, since the code evolved beyond the stage in the branch on another branch, I'll close this and reopen a new PR later.

**Pull request closed without merging**